File: server/nestjs/src/base/repository/product.repository.impl.ts
// src/product/product.service.ts

import { BadRequestException, Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Product } from 'base/entity/product.entity';
import { Repository } from 'typeorm';
import { ProductRepository } from './product.repository';
import { ProductRIVO, ProductROVO } from './vos';


@Injectable()
export class ProductRepositoryImpl implements ProductRepository{
  constructor(
    @InjectRepository(Product)
    private productDAO: Repository<Product>,
  ) {}

  async getAllProducts(): Promise<ProductROVO[]> {
    return this.productDAO.find();
  }

  async getProduct(id: number): Promise<ProductROVO> {
    return this.productDAO.findOne(id);
  }

  async createProduct(productData: Partial<ProductRIVO>): Promise<ProductROVO> {
    //TODO: id from request body should be removed nestjs problem - Security Risk
    //TODO: remove this console.log 
    console.log('productData', productData);
    const product = this.productDAO.create({name: productData.name, price: productData.price});
    return this.productDAO.save(product);
  }

  async updateProduct(id: number, productData: Partial<ProductRIVO>): Promise<ProductROVO> {
    await this.productDAO.update(id, productData);
    return this.productDAO.findOne(id);
  }

  async deleteProduct(id: number): Promise<void> {
    await this.productDAO.delete(id);
  }

  // async searchProductByNameAndPrice(name: string, price: number): Promise<ProductROVO[]> {
  //   return this.productDAO
  //     .createQueryBuilder('product')
  //     .where('product.name = :name', { name })
  //     .andWhere('product.price = :price', { price })
  //     .getMany();
  // }

  searchProductByNameAndPrice(name?: string, price?: number): Promise<Product[]> {
    if (!name && !price) {
      throw new BadRequestException('Either name or price must be provided');
    }
  
    if (name && price) {
      // search by name and price
      return this.productDAO.find({ name, price });
    } else if (name) {
      // search by name
      return this.productDAO.find({ name });
    } else {
      // search by price
      return this.productDAO.find({ price });
    }
  }
}

File: server/spring-kotlin-mm/app/src/main/kotlin/spring/kotlin/mm/product/repository/ProductRepositoryImpl.kt
package spring.kotlin.mm.product.repository


import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import spring.kotlin.mm.product.dao.ProductDaoRepository
import spring.kotlin.mm.product.entity.ProductEntity
import spring.kotlin.mm.product.repository.vo.ProductRIVO
import spring.kotlin.mm.product.repository.vo.ProductROVO

@Service
class ProductRepositoryImpl @Autowired constructor (private val productDAO: ProductDaoRepository) :
    ProductRepository {

    override fun save(product: ProductRIVO): ProductROVO {
        val productEntity = productDAO.save(ProductEntity(product.id, product.name, product.price))
        return ProductROVO(productEntity.id, productEntity.name, productEntity.price)
    }

    override fun deleteById(id: Long) {
        productDAO.deleteById(id)
    }

    override fun findById(id: Long): ProductROVO {
        val productEntity = productDAO.findById(id)
        if(productEntity.isEmpty) {
            throw NoSuchElementException("Product not found")
        }
        val entity = productEntity.get()
        return ProductROVO(entity.id, entity.name, entity.price)
    }

    override fun findAll(): List<ProductROVO> {
        return productDAO.findAll().map { ProductROVO(it.id, it.name, it.price) }
    }

    override fun findByName(name: String): List<ProductROVO> {
        return productDAO.findByName(name).map { ProductROVO(it.id, it.name, it.price) }
    }
}